### PR TITLE
Replace llvm::MD5 with SipHash-2-4

### PR DESCRIPTION
**Pull request by CodaFi**

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci test compiler performance

File: include/swift/Basic/StableHasher.h [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+#include "swift/Basic/StableHasher.h"
+#include "llvm/Support/Endian.h"
+#include <algorithm>
+
+namespace swift {
+
+/// A \c StableHasher is an implementation of a 128-bit stable hash - built for
+/// speed.
+///
+/// A "stable" hash in this context refers to the idea that the output of this
```
Comment (gottesmm) [inline include/swift/Basic/StableHasher.h:35]: Does LLVM already have a hash like this? Should this be done upstream?

Comment (CodaFi) [reply include/swift/Basic/StableHasher.h:35]: Clang has an implementation of SipHash hidden behind the ptrauth stuff that does not present the correct API surface. That's the most likely place to modify should we wish to upstream.

**Review by davidungar — COMMENTED** [review]
The code looks fine. I wonder how can we test or reassure ourselves that:
1. The new hash has similar or better collision performance than the old, and 
2. The implementation of the new algorithm computes the correct values?

**Comment by CodaFi** [comment]
@swift-ci test compiler performance

**Comment by CodaFi** [comment]
@swift-ci test compiler performance

**Comment by CodaFi** [comment]
@swift-ci test compiler performance

**Comment by CodaFi** [comment]
@swift-ci test compiler performance

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
@swift-ci smoke test

File: include/swift/Basic/StableHasher.h (hunk 1 above)
Comment (rjmccall) [reply include/swift/Basic/StableHasher.h:38]: You're more than welcome to extend the API surface.

**Comment by CodaFi** [comment]
@swift-ci smoke test Windows platform

**Comment by CodaFi** [comment]
@swift-ci smoke test Windows platform

**Comment by CodaFi** [comment]
@swift-ci smoke test Windows platform

**Comment by CodaFi** [comment]
@swift-ci smoke test Windows

**Comment by CodaFi** [comment]
@swift-ci smoke test Windows platform

**Comment by CodaFi** [comment]
@swift-ci smoke test Windows

**Comment by CodaFi** [comment]
@swift-ci smoke test

**Comment by CodaFi** [comment]
⛵

**Pull request merged**